test/ret_web/channels/entity_test.exs

```diff
@@ -0,0 +1,143 @@
+        assert update_message === update_from_payload["update_message"]
+      end)
+    end
+
+    test "save_entity_state is denied if not logged in", %{
+      socket: socket,
+      hub: hub
+    } do
+      {:ok, _, socket} = subscribe_and_join(socket, "hub:#{hub.hub_sid}", @default_join_params)
+
+      assert_reply push(socket, "save_entity_state", payload(:save_entity_state)), :error, %{
+        reason: :not_logged_in
+      }
+    end
+
+    test "save_entity_state is denied if account lacks pin permissions", %{
+      socket: socket,
+      hub: hub,
+      account: account
+    } do
+      {:ok, _, socket} =
+        subscribe_and_join(socket, "hub:#{hub.hub_sid}", join_params_for_account(account))
+
+      assert_reply push(socket, "save_entity_state", payload(:save_entity_state)), :error, %{
+        reason: :unauthorized
+      }
+    end
+  end
+
+  test "update_entity_state overwrites previous entity state", %{
+    socket: socket,
+    hub: hub,
+    account: account
+  } do
+    %HubRoleMembership{hub: hub, account: account} |> Repo.insert!()
+
+    {:ok, _, socket} =
+      subscribe_and_join(socket, "hub:#{hub.hub_sid}", join_params_for_account(account))
+
+    push(socket, "save_entity_state", payload(:save_entity_state))
+    assert_reply push(socket, "list_entities", %{}), :ok, %{data: [entity_state]}
+    # The page starts at 2
+    2 = Enum.at(entity_state.update_messages, 1)["data"]["networked-pdf"]["data"]["pageNumber"]
+
+    push(socket, "update_entity_state", payload(:update_entity_state))
+    assert_reply push(socket, "list_entities", %{}), :ok, %{data: [entity_state]}
+    # The page was updated to 1
+    assert Enum.at(entity_state.update_messages, 1)["data"]["networked-pdf"]["data"]["pageNumber"] ===
+             1
+  end
+
+  test "delete_entity_state deletes the entity with the matching nid", %{
+    socket: socket,
+    hub: hub,
+    account: account
+  } do
+    %HubRoleMembership{hub: hub, account: account} |> Repo.insert!()
+
+    {:ok, _, socket} =
+      subscribe_and_join(socket, "hub:#{hub.hub_sid}", join_params_for_account(account))
+
+    push(socket, "save_entity_state", payload(:save_entity_state))
+    push(socket, "save_entity_state", payload(:save_entity_state_2))
+    assert_reply push(socket, "delete_entity_state", payload(:delete_entity_state)), :ok
+    assert_reply push(socket, "list_entities", %{}), :ok, %{data: [entity]}
+    refute "ji5uv3q" === entity.create_message["networkedId"]
+    assert entity.create_message["networkId"] === payload(:save_entity_state_2)["nid"]
+    assert entity.create_message === payload(:save_entity_state_2)["create_message"]
+  end
+
+  defp local_file(filename) do
+    Path.join(Path.dirname(__ENV__.file), filename)
+  end
+
+  defp read_json(filename) do
+    Jason.decode!(File.read!(local_file(filename)))
+  end
+
+  defp payload(:save_entity_state) do
+    read_json("save_entity_state_payload.json")
```

Comment:
**nit:** This can be done at compile time, so we don’t have to perform a costly IO call every time this is invoked